ZenPacks/zenoss/OpenStackInfrastructure/apiclients/txapiclient.py

```diff
@@ -0,0 +1,839 @@
+
+        return self._apis[name]
+
+    def __getitem__(self, name):
+        return getattr(self, name)
+
+    def __call__(self, data=None, params=None, **kwargs):
+        # update url based on kwargs
+        if '/flavors/detail' in self.path and 'is_public' in kwargs:
+            if '?is_public=' in self.path:
+                self.path = self.path[:self.path.index('?is_public=')]
+            self.path = self.path + '?is_public=%s' % kwargs['is_public']
+
+        elif '/os-hypervisors' in self.path and \
+                'hypervisor_match' in kwargs and 'servers' in kwargs:
+            target = 'servers' if kwargs['servers'] else 'search'
+            self.path = '/os-hypervisors/%s/%s' % \
+                        (urllib.quote(kwargs['hypervisor_match'], safe=''),
+                         target)
+
+        elif '/os-hypervisors' in self.path and \
+                        'hypervisor_id' in kwargs:
+            self.path = '/os-hypervisors/%s' % kwargs['hypervisor_id']
+
+        elif '/cinderquotas' in self.path:
+            self.path = '/os-quota-sets/%s?usage=%s' % (kwargs['tenant'], kwargs['usage'])
+
+        # append meter name and query for ceilometer statistics
+        elif '/v2/meters/' in self.path and 'meter_name' in kwargs and 'queries' in kwargs:
+            self.path = self.path[:self.path.index('/v2/meters/') + len('/v2/meters/')]
+            self.path = self.path + kwargs['meter_name'] + '/statistics'
+
+        # append query for ceilometer samples
+        elif '/v2/samples' in self.path and 'queries' in kwargs:
+            self.path = self.path[:self.path.index('/v2/samples') + len('/v2/samples')]
+
+        # append query for ceilometer events
+        elif '/v2/events' in self.path and 'queries' in kwargs:
+            self.path = self.path[:self.path.index('/v2/events') + len('/v2/events')]
+
+        # append query for ceilometer alarms
+        elif '/v2/alarms' in self.path and 'queries' in kwargs:
+            self.path = self.path[:self.path.index('/v2/alarms') + len('/v2/alarms')]
+
+        # if there are query entries, add them to the path
+        if kwargs.has_key('queries'):
+            for query in kwargs['queries']:
+                self.path += '?q.field=%s&q.op=%s&q.type=%s&q.value=%s' % \
+                        (query.get('field',''), query.get('op',''),
+                         query.get('type',''), query.get('value',''))
+
+        return self.client.api_call(
+            self.path.encode('ascii', 'ignore'), data=data, params=params, **kwargs)
+
+class APIClientError(Exception):
+    """Parent class of all exceptions raised by api clients."""
+    pass
+
+class BadRequestError(APIClientError):
+    """Wrapper for HTTP 400 Bad Request error."""
+    pass
+
+
+class UnauthorizedError(APIClientError):
+    """Wrapper for HTTP 401 Unauthorized error."""
+    pass
+
+
+class NotFoundError(APIClientError):
+    """Wrapper for HTTP 400 Bad Request error."""
+    pass
+
+
+@inlineCallbacks
```

Comment:
yes, we need to have some unit tests for twisted api client. we just need a mock api server somewhere